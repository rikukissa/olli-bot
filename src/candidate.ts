const snowball = require("node-snowball");

export type Model = [
  string[],
  {
    [key: number]: Model;
  }
];

export type Candidate = string;
export type Candidates = Candidate[];
export type Replies = string[];

function randomFromArray(arr: Array<any>, amount: number) {
  if (arr.length === 0) {
    return [];
  }
  return Array(amount)
    .fill(null)
    .map(() => arr[Math.floor((arr.length - 1) * Math.random())]);
}

function getWords(message: string): string[] {
  return snowball.stemword(
    message
      .replace(/[\!\?\.]/g, "")
      .split(" ")
      .map(word => word.toLowerCase().trim()),
    "finnish"
  );
}

function getCandidateWithWeighting(candidates: Candidates, value: number) {
  const totalSlices = candidates.reduce((s, _, i) => s + i + 1, 0);
  const portions = candidates.map(
    (_, i) => (candidates.length - i) / totalSlices
  );

  let i = 0;
  let totalSum = 0;
  while (i < portions.length) {
    if (totalSum + portions[i] >= value || i === portions.length - 1) {
      break;
    }
    totalSum += portions[i];
    i++;
  }

  return candidates[i];
}

export function getRandomReplies(replies: Replies) {
  return randomFromArray(replies, 5);
}

export function getBestMatchingCandidate(
  model: Model,
  replies: Replies,
  message: string
) {
  const candidates = getCandidates(model, message, replies);

  const random = Math.random();

  return getCandidateWithWeighting(candidates, random);
}

export function storeCandidate(
  model: Model,
  message: string,
  candidate: Candidate
) {
  const words = getWords(message);

  let node = model;

  for (let word of words) {
    if (!node[1][word]) {
      node[1][word] = [[], {}];
    }
    node = node[1][word];
  }

  const matches = node[0];

  const indexInMatches = matches.indexOf(candidate);
  const existsAndNotMostPopular = indexInMatches > 0;

  if (existsAndNotMostPopular) {
    let betterOne = matches[indexInMatches - 1];
    matches[indexInMatches - 1] = candidate;
    matches[indexInMatches] = betterOne;
  }

  if (indexInMatches === -1) {
    matches.push(candidate);
  }
  return model;
}

function getNodeWithPath(model: Model, words: string[]): Model | undefined {
  let node = model;

  for (let word of words) {
    node = node[1][word];
    if (!node) {
      break;
    }
  }

  return node;
}

function getRepliesFromPath(model: Model, words: string[]) {
  let node = model;
  let i = 0;
  let results: string[] = [];

  while (node[1][words[i]] !== undefined) {
    node = node[1][words[i]];
    results = results.concat(node[0]);
    i++;
  }
  return results;
}

function getMatchesFromSiblings(tree: Model, parentNode: Model) {
  return Object.keys(parentNode[1])
    .filter(key => parentNode[1][key] !== tree)
    .reduce((memo, key) => memo.concat(parentNode[1][key][0]), []);
}

function getAllReplies(tree: Model): string[] {
  return Object.keys(tree[1]).reduce(
    (memo, key) => memo.concat(getAllReplies(tree[1][key])),
    tree[0]
  );
}

export function getCandidates(
  model: Model,
  message: string,
  replies: Replies
): string[] {
  const words = getWords(message);

  const matchingNode = getNodeWithPath(model, words);
  const parentNode = getNodeWithPath(model, words.slice(0, -1));

  const perfectMatches = matchingNode ? matchingNode[0] : [];

  const matchesFromFurtherInTree = matchingNode
    ? getAllReplies(matchingNode)
    : [];

  const matchesFromSiblings =
    matchingNode && parentNode
      ? getMatchesFromSiblings(matchingNode, parentNode)
      : [];

  const matchesFromBranchesBelow =
    words.length > 1
      ? getRepliesFromPath(model, words.slice(0, -1)).reverse()
      : [];

  let i = 0;
  let matchesFromSiblingsBelow: string[] = [];
  while (i < words.length) {
    const parent = getNodeWithPath(model, words.slice(0, words.length - i));
    if (parent) {
      matchesFromSiblingsBelow = getAllReplies(parent);
    }
    i++;
  }

  return [
    ...perfectMatches,
    ...matchesFromFurtherInTree,
    ...matchesFromSiblings,
    ...matchesFromBranchesBelow,
    ...matchesFromSiblingsBelow,
    ...getRandomReplies(replies)
  ];
}
