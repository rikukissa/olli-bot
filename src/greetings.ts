export const greetings = [
  'moro',
  'morjesta',
  'haudihou',
  'terve',
  'moikka',
  'moikkis',
  'moikkuli',
  'moro moro',
  'morjes',
  'morjens',
  'morjenttes',
  'morommoro',
  'moooooiiii',
  'moooiii',
  'mmoi!',
  'mooi',
  'yo yo',
  'tsau',
  'hellou',
];

export const greetingTo = [
  'peelo',
  'kaveri',
  'pelle xD',
  'kamu',
  'sä',
  'teikälle',
  'sälli',
  'kamuli',
  '',
  '',
  '',
  '',
  '',
];

export const emojis = [
  '😄',
  '😘',
  '😆',
  '😀',
  '😎',
  '💋',
  '😬',
  '😑',
  '😟',
  '😕',
  '',
  '',
  '',
  '',
  '',
];
