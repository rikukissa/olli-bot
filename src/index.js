const TelegramBot = require("node-telegram-bot-api");
const fs = require("fs");
const {
  loadFile,
  getRandomReplies,
  storeCandidate,
  persistTree,
  getCandidates,
  getReply
} = require("./train");
// replace the value below with the Telegram token you receive from @BotFather
const token = "";

// Create a bot that uses 'polling' to fetch new updates
const bot = new TelegramBot(token, { polling: true });

async function askForCandidate(fromId, candidates) {
  const options = candidates.map((candidate, i) => [
    {
      text: candidate,
      callback_data: i
    }
  ]);

  const opts = {
    reply_markup: {
      inline_keyboard: [
        ...options,
        [{ text: "Get a new random selection", callback_data: "r" }]
      ]
    }
  };

  const sentMessage = await bot.sendMessage(
    fromId,
    "Mikä olis bestest vastaus?",
    opts
  );

  return new Promise(resolve => {
    const listener = async callbackQuery => {
      await bot.deleteMessage(fromId, sentMessage.message_id);
      const input = callbackQuery.data.trim().toLowerCase();

      if (input === "r") {
        bot.removeListener("callback_query", listener);
        const candi = await askForCandidate(fromId, getRandomReplies());
        return resolve(candi);
      }

      const number = parseInt(input, 10);
      bot.sendMessage(
        fromId,
        "Gotcha 👍 " + "tästä eteenpäin tohon vastataan: " + candidates[number]
      );
      bot.removeListener("callback_query", listener);
      resolve(candidates[number]);
    };
    bot.on("callback_query", listener);
  });
}

function getMessage() {
  return new Promise(resolve => {
    const listener = msg => {
      const hasPrefix = msg.text.indexOf("!olli ") === 0;
      const privateChat = msg.chat.type === "private" && !hasPrefix;

      if (privateChat || hasPrefix) {
        resolve({
          message: msg.text.replace("!olli ", ""),
          from: msg.chat.id,
          private: privateChat
        });
        bot.removeListener(listener);
      }
    };
    bot.on("message", listener);
  });
}

async function loop(tree) {
  const { message, from, private } = await getMessage();

  if (private) {
    await bot.sendMessage(from, getReply(tree, message));

    loop(tree);
    return;
  }

  const candidates = getCandidates(tree, message);

  await bot.sendMessage(from, candidates[0]);
  const candidate = await askForCandidate(from, candidates);

  const newTree = storeCandidate(tree, message, candidate);

  persistTree(newTree);
  loop(newTree);
}

loop(loadFile());
