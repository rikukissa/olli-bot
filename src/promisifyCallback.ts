type Callback<T> = (data: T) => void;

export function promisifyCallback<T>(
  fn: (event: string, callback: Callback<T>) => void
) {
  return (event: string): Promise<T> => {
    return new Promise(resolve => {
      fn(event, (data: T) => resolve(data));
    });
  };
}
