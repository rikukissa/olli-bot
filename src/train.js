const fs = require("fs");
const path = require("path");
const snowball = require("node-snowball");

const replies = require("../replies.json");
const MODEL_PATH = path.join(__dirname, "./model.json");

function loadFile() {
  if (!fs.existsSync(MODEL_PATH)) {
    console.log("Creating a new model");

    return [[], {}];
  }
  return require(MODEL_PATH);
}
module.exports.loadFile = loadFile;

function persistTree(tree) {
  fs.writeFileSync(MODEL_PATH, JSON.stringify(tree, null, 2));
}
module.exports.persistTree = persistTree;

function randomFromArray(arr, amount) {
  if (arr.length === 0) {
    return [];
  }
  return Array(amount)
    .fill(null)
    .map(() => arr[Math.floor((arr.length - 1) * Math.random())]);
}

function getRandomReplies() {
  return randomFromArray(replies, 5);
}
module.exports.getRandomReplies = getRandomReplies;

function unique(arr) {
  return arr.filter((item, i) => arr.slice(0, i).indexOf(item) === -1);
}

function getWords(message) {
  return snowball.stemword(
    message
      .replace(/[\!\?\.]/g, "")
      .split(" ")
      .map(word => word.toLowerCase().trim()),
    "finnish"
  );
}

function storeCandidate(tree, message, candidate) {
  const words = getWords(message);

  let node = tree;

  for (word of words) {
    if (!node[1][word]) {
      node[1][word] = [[], {}];
    }
    node = node[1][word];
  }

  const matches = node[0];

  const indexInMatches = matches.indexOf(candidate);
  const existsAndNotMostPopular = indexInMatches > 0;

  if (existsAndNotMostPopular) {
    let betterOne = matches[indexInMatches - 1];
    list[indexInMatches - 1] = candidate;
    list[indexInMatches] = betterOne;
  }

  if (indexInMatches === -1) {
    matches.push(candidate);
  }
  return tree;
}

module.exports.storeCandidate = storeCandidate;

function getCandidates(tree, message) {
  const words = getWords(message);

  let node = tree;
  let i = 0;

  while (node[1][words[i]] !== undefined) {
    node = node[1][words[i]];
    i++;
  }

  const matches = node[0];

  if (matches.length === 0) {
    return getRandomReplies();
  }

  return [
    ...matches.slice(0, 2),
    ...unique(randomFromArray(matches.slice(2), 2)),
    ...getRandomReplies()
  ].slice(0, 5);
}

module.exports.getCandidates = getCandidates;
