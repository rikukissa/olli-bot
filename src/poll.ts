import * as TelegramBot from 'node-telegram-bot-api';
import { Candidates, Replies, Candidate, getRandomReplies } from './candidate';

type Options = Array<{ text: string; points: number }>;

const POLL_TIME = parseInt(process.env.POLL_TIME as string, 10);

export type MessageWithText = TelegramBot.Message & { text: string };

export function isTrainingMessage(message: MessageWithText) {
  return message.text.indexOf('!olli ') === 0;
}

export function removeTrainingCommandPrefix(text: string) {
  return text.replace('!olli ', '');
}

function createPollButtons(options: Options) {
  return {
    inline_keyboard: [
      ...options.map(({ text, points }, i) => [
        {
          text: `${text} ${points > 0 ? `(${points})` : ''}`,
          callback_data: i.toString(),
        },
      ]),
      MORE_BUTTON,
    ],
  };
}

const MORE_BUTTON = [{ text: '➕ Lisää', callback_data: 'more' }];
export async function pollForBestCandidate(
  bot: TelegramBot,
  chatId: number,
  candidates: Candidates,
  replies: Replies,
): Promise<Candidate> {
  // Value changes every time a candidate is selected by someone
  let pollOptions = candidates.map(text => ({ text, points: 0 }));

  const pollButtons = createPollButtons(pollOptions);

  const getPollTitle = (timeLeft: number) =>
    `Mikä olis bestest vastaus? (${Math.floor(timeLeft / 1000)}s vastausaikaa)`;

  const titleMessage = await bot.sendMessage(chatId, getPollTitle(POLL_TIME));
  const buttonMessage = await bot.sendMessage(chatId, 'Vaihtoehdot:', {
    reply_markup: pollButtons,
  });

  const rerenderText = (timeLeft: number) =>
    bot.editMessageText(getPollTitle(timeLeft), {
      chat_id: chatId,
      message_id: titleMessage.message_id,
    });

  const rerenderButtons = () =>
    bot.editMessageReplyMarkup(createPollButtons(pollOptions), {
      chat_id: chatId,
      message_id: buttonMessage.message_id,
    });

  return new Promise<string>(resolve => {
    function clear() {
      bot.removeListener('callback_query', listener);
      return Promise.all([
        bot.deleteMessage(chatId, titleMessage.message_id.toString()),
        bot.deleteMessage(chatId, buttonMessage.message_id.toString()),
      ]);
    }

    async function onceASecond(timeLeft: number) {
      if (timeLeft <= 0) {
        await clear();

        const topOption = pollOptions.reduce(
          (best, option) => (option.points > best.points ? option : best),
        );
        const topScore = topOption.points;

        const allWithTopScore = pollOptions.filter(
          ({ points }) => points === topScore,
        );
        const result =
          allWithTopScore[Math.floor(Math.random() * allWithTopScore.length)];

        return resolve(result.text);
      }

      await rerenderText(timeLeft - 1000);
      setTimeout(() => onceASecond(timeLeft - 1000), 1000);
    }

    let firstAnswerReceived = false;

    const listener = async (callbackQuery: TelegramBot.CallbackQuery) => {
      if (chatId !== (callbackQuery.message && callbackQuery.message.chat.id)) {
        return;
      }

      const selectedOptionData = callbackQuery.data as string;

      if (selectedOptionData === 'more') {
        pollOptions = pollOptions.concat(
          getRandomReplies(replies).map(text => ({ text, points: 0 })),
        );
      } else {
        const num = parseInt(selectedOptionData, 10);
        pollOptions[num].points++;

        if (!firstAnswerReceived) {
          firstAnswerReceived = true;
          onceASecond(POLL_TIME);
        }
      }

      await rerenderButtons();
    };

    bot.on('callback_query', listener);
  });
}
