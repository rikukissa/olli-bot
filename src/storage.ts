import * as Octo from "@octokit/rest";
import { Model } from "./candidate";

const octokit = new Octo();

const GIST_ID = process.env.GIST_ID as string;
const REPLIES_GIST_ID = process.env.REPLIES_GIST_ID as string;
const GITHUB_TOKEN = process.env.GITHUB_TOKEN as string;

const authPromise = octokit.authenticate({
  type: "token",
  token: GITHUB_TOKEN
});

export async function getReplies() {
  await authPromise;

  const repliesGist = await octokit.gists.get({
    gist_id: REPLIES_GIST_ID
  });

  return JSON.parse(repliesGist.data.files["replies.json"].content);
}

export async function getModel() {
  await authPromise;

  const modelGist = await octokit.gists.get({
    gist_id: GIST_ID
  });

  return JSON.parse(modelGist.data.files["model.json"].content);
}

export async function storeModel(model: Model) {
  await authPromise;

  return octokit.gists.edit({
    gist_id: GIST_ID,
    files: {
      "model.json": {
        filename: "model.json",
        content: JSON.stringify(model, null, 2)
      }
    } as any // TODO
  });
}
