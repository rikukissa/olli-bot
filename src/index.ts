import * as dotenv from 'dotenv';
dotenv.config();

import * as TelegramBot from 'node-telegram-bot-api';
import * as Levenshtein from 'levenshtein';
import { merge } from 'lodash';
import { greetings, emojis, greetingTo } from './greetings';

import {
  storeCandidate,
  getCandidates,
  getBestMatchingCandidate,
  Replies,
  Model,
} from './candidate';
import { storeModel, getReplies, getModel } from './storage';
import {
  pollForBestCandidate,
  MessageWithText,
  isTrainingMessage,
  removeTrainingCommandPrefix,
} from './poll';

const TELEGRAM_TOKEN = process.env.TELEGRAM_TOKEN as string;

const bot = new TelegramBot(TELEGRAM_TOKEN, { polling: true });

function waitForChannelMessage(chatId: number) {
  return new Promise<TelegramBot.Message>(resolve => {
    const handler = (message: TelegramBot.Message) => {
      if (message.chat.id === chatId) {
        bot.removeListener('message', handler);
        resolve(message);
      }
    };
    bot.on('message', handler);
  });
}

function wait(time: number) {
  return new Promise(resolve => setTimeout(resolve, time));
}

function resemblesMeaningOfLife(message: string): boolean {
  const { distance } = new Levenshtein(message, 'mikä on elämän tarkoitus');

  return distance < 3;
}
function resemblesGreetings(message: string): string[] {
  return greetings.filter(greeting => {
    const { distance } = new Levenshtein(message, greeting);
    return distance < 2;
  });
}

function constructGreeting() {
  const greeting = greetings[Math.floor(Math.random() * greetings.length)];
  const to = greetingTo[Math.floor(Math.random() * greetingTo.length)];
  const emoji = emojis[Math.floor(Math.random() * emojis.length)];

  if (to === '' && emoji === '') {
    return `${greeting}!!`;
  }

  if (to === '') {
    return `${greeting} ${emoji}${emoji}`;
  }

  return `${greeting} ${to} ${emoji}`;
}

async function sendBestMatchingCandidate(
  model: Model,
  replies: Replies,
  message: MessageWithText,
  chatId: number,
): Promise<Model> {
  await wait(Math.random() * 2000);
  await bot.sendChatAction(chatId, 'typing');

  const matchingGreetings = resemblesGreetings(message.text);

  if (matchingGreetings.length > 0) {
    const randomizedGreeting = constructGreeting();

    await wait(Math.random() * 4000 + randomizedGreeting.length * 300);
    await bot.sendMessage(chatId, randomizedGreeting);

    return model;
  }

  if (resemblesMeaningOfLife(message.text)) {
    await bot.sendMessage(chatId, '42');
    await wait(1800);
    await bot.sendMessage(chatId, 'ja rööki :D');

    return model;
  }

  const candidate = getBestMatchingCandidate(model, replies, message.text);
  await wait(Math.random() * 4000 + candidate.length * 300);

  await bot.sendMessage(chatId, candidate);
  return model;
}

async function handleMessage(
  model: Model,
  replies: Replies,
  message: MessageWithText,
  chatId: number,
): Promise<Model> {
  if (!message.text) {
    return model;
  }

  if (!isTrainingMessage(message)) {
    await sendBestMatchingCandidate(model, replies, message, chatId);
    return model;
  }

  const text = removeTrainingCommandPrefix(message.text);

  const candidates = getCandidates(model, text, replies).slice(0, 5);
  const candidateWithHighestScore = candidates[0];

  const candidate = await pollForBestCandidate(
    bot,
    chatId,
    candidates,
    replies,
  );

  if (candidateWithHighestScore !== candidate) {
    await bot.sendMessage(chatId, candidate);
  }

  return storeCandidate(model, text, candidate);
}

async function runBot() {
  let model = await getModel();
  const replies = await getReplies();
  const knownChannels: number[] = [];

  // Channel specific "process"
  async function messageLoop(chatId: number, message?: TelegramBot.Message) {
    const receivedMessage = message || (await waitForChannelMessage(chatId));

    if (!receivedMessage.text) {
      messageLoop(chatId);
      return;
    }
    const newModel = await handleMessage(
      model,
      replies,
      receivedMessage as MessageWithText,
      chatId,
    );

    // Many running chats can modify the "global" messages model
    model = merge(model, newModel);

    storeModel(model);
    messageLoop(chatId);
  }

  bot.on('message', (message: TelegramBot.Message) => {
    const channelLoopAlreadyRunning =
      knownChannels.indexOf(message.chat.id) > -1;

    if (channelLoopAlreadyRunning) {
      return;
    }

    knownChannels.push(message.chat.id);

    if (message.text === '/start') {
      messageLoop(message.chat.id);
      return;
    }

    messageLoop(message.chat.id, message);
  });
}

runBot();
