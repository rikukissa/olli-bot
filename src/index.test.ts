import { getBestMatchingCandidate } from "src/candidate";
const model = require("./test/model.json");
const replies = require("./test/replies.json");

const mockMath = Object.create(global.Math);
mockMath.random = () => 0;
global.Math = mockMath;

describe("replies", () => {
  it("foobar", () => {
    expect(getBestMatchingCandidate(model, replies, "moro")).toMatchSnapshot();
    expect(
      getBestMatchingCandidate(model, replies, "mit√§s kekke")
    ).toMatchSnapshot();
  });
});
