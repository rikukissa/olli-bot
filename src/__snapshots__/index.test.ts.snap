// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replies foobar 1`] = `"moro boijaat"`;

exports[`replies foobar 2`] = `"laitoin pari biisi√§ poimintoihin"`;
